src/retry.rs

```diff
@@ -0,0 +1,438 @@
+//! the default, and you will need to explicitly include that behavior.
+//!
+//! All policies default to including a retry budget that permits 20% extra
+//! requests to be sent.
+//!
+//! # Scoped
+//!
+//! A client's retry policy is scoped. That means that the policy doesn't
+//! apply to all requests, but only those within a user-defined scope.
+//!
+//! Since all policies include a budget by default, it doesn't make sense to
+//! apply it on _all_ requests. Rather, the retry history applied by a budget
+//! should likely only be applied to the same host.
+//!
+//! # Classifiers
+//!
+//! A retry policy needs to be configured with a classifier that determines
+//! if a request should be retried. Knowledge of the destination server's
+//! behavior is required to make a safe classifier. **Requests should not be
+//! retried** if the server cannot safely handle the same request twice, or if
+//! it causes side effects.
+//!
+//! Some common properties to check include if the request method is
+//! idempotent, or if the response status code indicates a transient error.
+
+use std::sync::Arc;
+use std::time::Duration;
+
+use tower::retry::budget::{Budget as _, TpsBudget as Budget};
+
+/// Builder to configure retries
+///
+/// Construct with [`for_host()`].
+#[derive(Debug)]
+pub struct Builder {
+    //backoff: Backoff,
+    budget: Option<f32>,
+    classifier: classify::Classifier,
+    max_retries_per_request: u32,
+    scope: scope::Scoped,
+}
+
+/// The internal type that we convert the builder into, that implements
+/// tower::retry::Policy privately.
+#[derive(Clone, Debug)]
+pub(crate) struct Policy {
+    budget: Option<Arc<Budget>>,
+    classifier: classify::Classifier,
+    max_retries_per_request: u32,
+    retry_cnt: u32,
+    scope: scope::Scoped,
+}
+
+//#[derive(Debug)]
+//struct Backoff;
+
+/// Create a retry builder with a request scope.
+///
+/// To provide a scope that isn't a closure, use the more general
+/// [`Builder::scoped()`].
+pub fn for_host<S>(host: S) -> Builder
+where
+    S: PartialEq<str> + Send + Sync + 'static,
+{
+    scoped(move |req| &host == req.uri().host().unwrap_or(""))
+}
+
+fn scoped<F>(func: F) -> Builder
+where
+    F: Fn(&Req) -> bool + Send + Sync + 'static,
+{
+    Builder::scoped(scope::ScopeFn(func))
+}
+
+// ===== impl Builder =====
+
+impl Builder {
+    /// Create a scoped retry policy.
+    ///
+    /// For a more convenient constructor, see [`for_host()`].
+    pub fn scoped(scope: impl scope::Scope) -> Self {
+        Self {
+            budget: Some(0.2),
+            // XXX: should the default be Never?
+            classifier: classify::Classifier::ProtocolNacks,
+            max_retries_per_request: 2, // on top of the original
+            scope: scope::Scoped::Dyn(Arc::new(scope)),
+        }
+    }
+
+    /// Set no retry budget.
+    pub fn no_budget(mut self) -> Self {
+        self.budget = None;
+        self
+    }
```

Comment:
Nit: It isn't clear if "no budget" is "never retry" (the budget is zero) or "always retry" (retries aren't restricted by a budget). I would recommend calling this something like `unlimited_retries` or `disable_budget`. (Or at the very least adding more explicit wording to the doc comment.)